modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/AclRoutingSearchProcessor.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.pipeline.common;
+
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.common.hash.MurmurHash3;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.index.query.TermsQueryBuilder;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.search.pipeline.AbstractProcessor;
+import org.opensearch.search.pipeline.Processor;
+import org.opensearch.search.pipeline.SearchRequestProcessor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Search processor that adds routing based on ACL fields in the query.
+ */
+public class AclRoutingSearchProcessor extends AbstractProcessor implements SearchRequestProcessor {
+
+    /**
+     * The type name for this processor.
+     */
+    public static final String TYPE = "acl_routing_search";
+
+    private final String aclField;
+    private final boolean extractFromQuery;
+
+    /**
+     * Constructor for AclRoutingSearchProcessor.
+     *
+     * @param tag processor tag
+     * @param description processor description
+     * @param ignoreFailure whether to ignore failures
+     * @param aclField the field to extract ACL values from
+     * @param extractFromQuery whether to extract ACL values from query
+     */
+    public AclRoutingSearchProcessor(String tag, String description, boolean ignoreFailure, String aclField, boolean extractFromQuery) {
+        super(tag, description, ignoreFailure);
+        this.aclField = aclField;
+        this.extractFromQuery = extractFromQuery;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public SearchRequest processRequest(SearchRequest request) throws Exception {
+        if (!extractFromQuery || request.source() == null) {
+            return request;
+        }
+
+        QueryBuilder query = request.source().query();
+        if (query == null) {
+            return request;
+        }
+
+        List<String> aclValues = extractAclValues(query);
+        if (aclValues.isEmpty()) {
+            return request;
+        }
+
+        // Generate routing values
+        String[] routingValues = aclValues.stream().map(this::generateRoutingValue).toArray(String[]::new);
+
+        // Set routing on the request
+        request.routing(routingValues);
+
+        return request;
+    }
+
+    private List<String> extractAclValues(QueryBuilder query) {
+        List<String> aclValues = new ArrayList<>();
+
+        if (query instanceof TermQueryBuilder) {
```

Comment:
Similar sort of comment as on the hierarchical processor regarding using `QueryBuilderVisitor`. Can we use that instead?